File: puzzles/haskell/mime-type.hs
import System.IO
import Control.Monad
import System.FilePath
import Data.Char
import qualified Data.Map as Map

main :: IO ()
main = do
  hSetBuffering stdout NoBuffering -- DO NOT REMOVE

  input_line <- getLine
  let nbElements = read input_line :: Int
  input_line <- getLine
  let nbNames = read input_line :: Int

  associationList <- replicateM nbElements $ do
    input_line <- getLine
    let input = words input_line
    let extension = input!!0
    let mimeType = input!!1
    return (map toLower extension, mimeType)

  let table = Map.fromList associationList

  replicateM nbNames $ do
    name <- getLine
    let extension = map toLower (takeExtension name)
    if extension == "" || extension == "." then
      putStrLn "UNKNOWN"
    else do
      let extWithoutDot = drop 1 extension
      case Map.lookup extWithoutDot table of
        Just mimeType -> putStrLn mimeType
        Nothing -> putStrLn "UNKNOWN"
    return ()
  return ()

File: puzzles/python2/mime_type.py
# dictionary to map file extensions to mime types
table = {}
nb_elements = int(raw_input())
nb_names = int(raw_input())
for _ in range(nb_elements):
    extension, mime_type = raw_input().split()
    table[extension.lower()] = mime_type

for _ in range(nb_names):
    name = raw_input().lower()
    dot_index = name.rfind(".")
    if dot_index == -1 or dot_index == len(name) - 1:
        print("UNKNOWN")
    else:
        extension = name[dot_index + 1:]
        try:
            print(table[extension])
        except KeyError:
            print("UNKNOWN")
